doc/memos/rdm-draft-buschsieweke-timer-api.md

```diff
@@ -0,0 +1,201 @@
+  RTC, the resolution can differ from 1 second. RTTs provide all mandatory
+  features of the [RTT API](http://api.riot-os.org/group__drivers__periph__rtt.html)
+- Software Timer: A mechanism to run code at a given point in time in the
+  future. The number of concurrently usable software timers is only limited by
+  their memory requirements and the load a system can handle.
+- System Time: Common time reference of the local device, usually as a single
+  integer counting the clock ticks passed since some reference point
+
+# 1. Introduction
+
+## Motivation
+
+Many applications and use cases for using timer hardware for various things
+ranging from measuring time, delaying the execution of a thread, or running
+code at specific points in time exists. Convenience, portability and ease of
+use are therefore essential for a user friendly embedded OS.
+
+Actual timer hardware however is difficult to use. One main reason is that
+timer hardware can differ significantly starting with the clock frequency,
+the timer width (8 bit, 16 bit, 24 bit, and 32 bit are all common), and their
+time representation. Another reason is that hardware timers only allow setting
+a fixed number of alarms - possible only one.
+
+A high level timer API is needed to provide convenient and portable access to
+the hardware features and is essential when multiple virtual timers need to
+be provided on top of a single hardware timer.
+
+## Related Timer APIs and Commonly Provided Features
+
+The following list of related APIs provided by other operating systems targeting
+constraint devices of class C2 according to RFC 7228 terminology has been
+studied:
+
+- FreeRTOS:
+    - [Software Timer Feature description](https://www.freertos.org/RTOS-software-timer.html)
+    - [Software Timer API reference](https://www.freertos.org/FreeRTOS-Software-Timer-API-Functions.html)
+    - [Task Control API reference](https://www.freertos.org/a00112.html)
+    - [`xTaskGetTickCount()` API reference](https://www.freertos.org/a00021.html#xTaskGetTickCount)
+- Zephyr:
+    - [Timer API documentation](https://docs.zephyrproject.org/1.12.0/kernel/timing/timers.html)
+    - [Kernel Clocks documentation](https://docs.zephyrproject.org/1.9.0/kernel/timing/clocks.html)
+- Mbed OS:
+    - [Timer API documentation](https://os.mbed.com/docs/mbed-os/v5.14/apis/timer.html)
+    - [Ticker API documentation](https://os.mbed.com/docs/mbed-os/v5.14/apis/ticker.html)
+    - [`wait()` API documentation](https://os.mbed.com/docs/mbed-os/v5.14/apis/wait.html)
+- Contiki
+    - [Wiki page on timers](https://github.com/contiki-os/contiki/wiki/Timers)
+- TinyOS:
+    - [Timer header file](https://github.com/tinyos/tinyos-main/blob/master/tos/lib/timer/Timer.nc)
+    - [TOSThreads Tutorial](http://tinyos.stanford.edu/tinyos-wiki/index.php/TOSThreads_Tutorial)
+- Mynewt:
+    - [OS Time API reference](https://mynewt.apache.org/latest/os/core_os/time/os_time.html)
+
```

Comment:
@maribu @kaspar030 Makes sense to add RIOT and xtimer here, does it not?

Reply:
I added xtimer. I guess it also makes sense to add power management as metric; but that will take some time to research for all the other OSs